packages/data/src/plugins/persistence/README.md

```diff
@@ -0,0 +1,44 @@
+Persistence Plugin
+==================
+
+The persistence plugin enhances a registry to enable registered stores to opt in to persistent storage.
+
+By default, persistence occurs by [`localStorage`](https://developer.mozilla.org/en-US/docs/Web/API/Window/localStorage). This can be changed using the [`setStorage` function](#api) defined on the plugin. Unless set otherwise, state will be persisted on the `WP_DATA` key in storage.
+
+## Usage
+
+Call the `use` method on the default or your own registry to include the persistence plugin:
+
+```js
+wp.data.use( 'persistence' );
```

Comment:
Can I ask why you preferred the "string" syntax over the "function" one (similar to redux).

I prefer personally `wp.data.use( createPersistencePlugin( {
   storage: window.localStorage,
   key: 'myKey' 
} ) )`

for these reasons:

 - No global state which means no conflict between registries
 - Ability to create custom plugins easily without a need for a `addPlugin`/`registerPlugin` API.

Reply:
It doesn't strictly need to be a string. It's this way largely as a convenience. It could be a pointer to the function as well:

```js
wp.data.use( wp.data.plugins.persistence );
```

Part of the thinking was consistency between plugin initialization. Considering others that we may explore to refactor as plugins (resolvers, controls), they don't require such "creation". 

So it would be odd to have both `wp.data.plugins.persistence.createPlugin` and then just `wp.data.plugins.resolvers` (or a `wp.data.plugins.resolvers.createPlugin` that accepted no arguments).

And there already is one layer of creation with the plugin called as a function with the registry as an argument. It's possible we could have options as an optional argument to the `use`, so we have something like:

```js
wp.data
    .use( wp.data.plugins.persistence, { storageKey: '...' } )
    .use( wp.data.plugins.resolvers );
```

>No global state which means no conflict between registries

It's still not very clear to me the best way we want these "default" plugins to be brought in. If we want this to occur within data itself, it seems reasonable that there's some global context; how else would we customize the storage behavior without the storage key (with user ID) being known to the data module (bad for generalism)?

I'd also thought it might be suitably explicit if the modules which register stores would always make sure to call `use` before calling to register their own store.

```js
use( persistence );

registerStore( 'core/edit-post', { /* ... */ } );
```

This would require some level of deduplication of plugins, though it's easy enough to achieve.

Reply:
`createPlugin` is the responsibility of the plugin author, I'm mostly concerned about reducing the API surface as much as we can. If we have just `registry.use( pluginFunction )` I'm fine with that.

I can't see the downside of this, it just allows everything:

 - configuration using `createPlugin` like function returning the plugin function
 - non-configurable plugins 

> It's still not very clear to me the best way we want these "default" plugins to be brought in. If we want this to occur within data itself, it seems reasonable that there's some global context; 

I think we should limit the global context as much as we can.
If we do:

```js
wp.data.use( wp.data.persistence.createPlugin( config ) )
```

in an inline script, we have both the defaut plugins built-in and also no globals and no leaking to other registries. If you want to customize the storage engine, you do it in that call above. Maybe I'm missing something here?

> I'd also thought it might be suitably explicit if the modules which register stores would always make sure to call use before calling to register their own store.

For me, that's probably the most compelling reason for the globally available plugins `use( 'persistence' )` but even though, I'm not certain we can make it work without having plugins declared by a store messing up with plugins declared by other stores.

Reply:
The main remaining piece is consistency between including multiple plugins. I would not like, for example:

```js
wp.data
   .use( wp.data.plugins.persistence.createPlugin( config ) );
   .use( wp.data.plugins.resolvers );
```

...because this breaks expectations on how a developer works with a specific plugin (need to understand whether it has a `createPlugin` or not).

That's where I'm thinking of either:

```js
wp.data
    .use( wp.data.plugins.persistence, { storageKey: '...' } )
    .use( wp.data.plugins.resolvers );
```

Where the object effectively causes the plugin to be "created" in the same way you've mentioned. This would look like:

```js
function plugin( registry, options ) {
   // ...initialize storage from options, available to closure
}
```

Alternatively, we could assume that `use` should be called with a created plugin, but have the developer pass the registry, in place of current implementation of `use` calling with `registry` as an argument automatically.

```js
wp.data
    .use( wp.data.plugins.persistence.create( registry, { storageKey: '...' } )
    .use( wp.data.plugins.resolvers.create( registry ) );
    // Or: With default registry?
    // .use( wp.data.plugins.resolvers.create() );
```

Reply:
> wp.data
    .use( wp.data.plugins.persistence, { storageKey: '...' } )
    .use( wp.data.plugins.resolvers );

This one has my vote then.

I'm thinking most of these make sense anyway, some small differences but in the end it's the same thing. So let's move forward with what you think is best.

Reply:
Shorter is better, second that 😃